// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless  by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// python -m grpc_tools.protoc -I=. --python_out=. --pyi_out=.  --grpc_python_out=. ./chat_app.proto

syntax = "proto3";

import "google/protobuf/timestamp.proto";


option java_multiple_files = true;
option java_package = "io.grpc.examples.chatAPP";
option java_outer_classname = "ChatAppProto";
option objc_class_prefix = "CAP";

package chatapp;

// Interface exported by the server.
service ChatApp {

  rpc Login(User) returns (RequestReply); 

  rpc SignUp(User) returns (RequestReply); 

  rpc ListAll(ListAllRequest) returns (UserList); 

  rpc DeleteUser(User) returns (RequestReply);

  rpc SendMessage(ChatMessage) returns (RequestReply);

  rpc ReceiveMessage(User) returns (stream ChatMessage);

}

message RequestReply {
     string reply = 1; 
     int32 request_status = 2;
}

message User {
    string username = 1; // Username (uniquely identifies user)
}

message UserList {
    repeated User users = 1;
}

message ListAllRequest {
    string username_filter = 1; 
}

message ChatMessage {
    User sender = 1;  // Sender user of the message.
    UserList destinataries = 2; // Destinatary users of the message.
    string text = 3; // Text body of the message.
    google.protobuf.Timestamp date = 4; // Datetime at which message is sent.
}


